p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,355 @@
+
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/pion/ice/v2"
+	pionlogger "github.com/pion/logging"
+	"github.com/pion/webrtc/v3"
+)
+
+type connMultiaddrs struct {
+	local, remote ma.Multiaddr
+}
+
+var _ network.ConnMultiaddrs = &connMultiaddrs{}
+
+func (c *connMultiaddrs) LocalMultiaddr() ma.Multiaddr  { return c.local }
+func (c *connMultiaddrs) RemoteMultiaddr() ma.Multiaddr { return c.remote }
+
+const (
+	candidateSetupTimeout         = 20 * time.Second
+	DefaultMaxInFlightConnections = 10
+)
+
+type candidateAddr struct {
+	ufrag string
+	raddr *net.UDPAddr
+}
+
+type listener struct {
+	transport *WebRTCTransport
+
+	mux ice.UDPMux
+
+	config                    webrtc.Configuration
+	localFingerprint          webrtc.DTLSFingerprint
+	localFingerprintMultibase string
+
+	localAddr      net.Addr
+	localMultiaddr ma.Multiaddr
+
+	// buffered incoming connections
+	acceptQueue chan tpt.CapableConn
+
+	// Accepting a connection requires instantiating a peerconnection
+	// and a noise connection which is expensive. We therefore limit
+	// the number of in-flight connection requests. A connection
+	// is considered to be in flight from the instant it is handled
+	// until it is dequeued by a call to Accept, or errors out in some
+	// way.
+	inFlightInputQueue chan struct{}
+
+	// used to control the lifecycle of the listener
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+var _ tpt.Listener = &listener{}
+
+func newListener(transport *WebRTCTransport, laddr ma.Multiaddr, socket net.PacketConn, config webrtc.Configuration) (*listener, error) {
+	localFingerprints, err := config.Certificates[0].GetFingerprints()
+	if err != nil {
+		return nil, err
+	}
+
+	localMh, err := hex.DecodeString(strings.ReplaceAll(localFingerprints[0].Value, ":", ""))
+	if err != nil {
+		return nil, err
+	}
+	localMhBuf, err := multihash.Encode(localMh, multihash.SHA2_256)
+	if err != nil {
+		return nil, err
+	}
+	localFpMultibase, err := multibase.Encode(multibase.Base64url, localMhBuf)
+	if err != nil {
+		return nil, err
+	}
+
+	inFlightQueueCh := make(chan struct{}, transport.maxInFlightConnections)
+	for i := uint32(0); i < transport.maxInFlightConnections; i++ {
+		inFlightQueueCh <- struct{}{}
+	}
+
+	l := &listener{
+		transport:                 transport,
+		config:                    config,
+		localFingerprint:          localFingerprints[0],
+		localFingerprintMultibase: localFpMultibase,
+		localMultiaddr:            laddr,
+		localAddr:                 socket.LocalAddr(),
+		acceptQueue:               make(chan tpt.CapableConn),
+		inFlightInputQueue:        inFlightQueueCh,
+	}
+
+	l.ctx, l.cancel = context.WithCancel(context.Background())
+	mux := udpmux.NewUDPMux(socket, func(ufrag string, addr net.Addr) bool {
```

Comment:
Totally agree, this is beyond messy. I'm wondering why we're not following a listener / accept pattern here. I believe this would also make it easier to resolve your comments around cleanup of the connection when the rcmgr blocks the incoming connection. I'll try to refactor this tomorrow. It's not trivial, but shouldn't be too difficult either.

Reply:
Done! It's so much cleaner!